### DISCUSSION TITLE
How to create a portable version of VS Code + Z Open Editor (Java runtime problem)

### DISCUSSION THREAD

**Discussion by FALLAI-Denis** [IBM/zopeneditor-about#381] [Q&A]
Hi,

I need to create a portable version of VS Code, for a Windows context, that includes all necessary extensions, including IBM Z Open Editor.
This portable version must be completely autonomous. This implies that third-party components are also embedded in the portable installation.

I created an execution context with the "zip" version of VS Code for Windows: https://code.visualstudio.com/sha/download?build=stable&os=win32-x64-archive
The deployment architecture of this archive is documented here: https://code.visualstudio.com/docs/editor/portable 

![image](https://github.com/IBM/zopeneditor-about/assets/63861965/0aab61f6-7218-4123-a28d-5dbef20c7d4c)

I installed the various extensions which were placed in the `data\extensions` subfolder.
User context and preferences are placed in the `data\user-data` subfolder.

IBM Z Open Editor requires a Java runtime that is not necessarily available (or not at the correct version) on the host machine. 
So I want to embed a JDK (or JRE) in the portable installation.
I copied a JDK into a `data\JAVA` subfolder:
```
X:\Installation_folder\data\JAVA\bin\java.exe --version
openjdk 11.0.16.1 2022-08-12
IBM Semeru Runtime Open Edition 11.0.16.1 (build 11.0.16.1+1)
Eclipse OpenJ9 VM 11.0.16.1 (build openj9-0.33.1, JRE 11 Windows 11 amd64-64-Bit Compressed References 20220818_429 (JIT enabled, AOT enabled)
OpenJ9   - 1d9d16830
OMR      - b58aa2708
JCL      - 84b84a0409 based on jdk-11.0.16.1+1)
```
This is the IBM Semeru Java version. 
On current test machine, the default Java runtime is the Microsoft OpenJDK:
```
java --version
openjdk 11.0.13 2021-10-19 LTS
OpenJDK Runtime Environment Microsoft-27990 (build 11.0.13+8-LTS)
OpenJDK 64-Bit Server VM Microsoft-27990 (build 11.0.13+8-LTS, mixed mode)
```

But I have to give the path to this JAVA runtime using the user `settings.json`, using the `zopeneditor.JAVA_HOME` parameter.
In settings.json files it is not possible to refer to environment variables: the `{$env:ENVAR}` syntax is not supported. 
You must use one of the VS Code predefined variables: https://code.visualstudio.com/docs/editor/variables-reference#_predefined-variables
The only available variable that refers to the VS Code installation context is `${execPath}`.
However, this variable references the VS Code executable itself, and not only the folder containing this executable.
To return to the path you must use the expression `${execPath}\..\`.

This syntax works well for referencing files. 
It also works with ZOE to detect the JAVA version used, but not to run the Language Server necessary for ZOE:

In User `settings.json`: `"zopeneditor.JAVA_HOME": "${execPath}\\..\\data\\JAVA"`

In output for extension:
```
2024-02-04T15:40:41.768+01:00 DEBUG: UsageReporter: Sending the following telemetry data: {"eventName":"java/found","properties":{"editor":"Visual Studio Code","platform":"win32","java.majorversion":"11","java.vendor":"ibm"},"measurements":{"sinceActivated":3.114}}
2024-02-04T15:40:41.775+01:00 DEBUG: JavaFinder: Used the "zopeneditor.JAVA_HOME" user setting to locate Java.
...
Is a valid Java configured: true
  - Java: ${execPath}\..\data\JAVA\bin\java.exe
  - Java memory allocation: 1024
  - Java startup user settings: Not defined
```
According ZOE log, JavaFinder find the IBM Semeru version for portable installation. 

When opening a COBOL file:
```
2024-02-04T15:49:41.615+01:00 DEBUG: ZLanguageServer.buildServerOptions: Starting COBOL language server from "language-server".
2024-02-04T15:49:41.676+01:00 DEBUG: ZLanguageServer: Will start a language server with command "${execPath}\..\data\JAVA\bin\java.exe -Declipse.application=com.ibm.wazi.lsp.cobol.core.languageServer,-Dosgi.bundles.defaultStartLevel=4,-Declipse.product=com.ibm.wazi.lsp.cobol.core.languageServerProduct,-Xmx1024m,-Xdisablejavadump,-jar,X:\installation_folder\data\extensions\ibm.zopeneditor-3.3.3\language-server\plugins\org.eclipse.equinox.launcher_1.6.400.v20210924-0641.jar,-configuration,X:\installation_folder\data\extensions\ibm.zopeneditor-3.3.3\language-server\cobol_config_win,-data,X:\installation_folder\data\tmp\_ws_1707058181675".
2024-02-04T15:49:41.685+01:00 DEBUG: AbstractZLanguageClient.initializeLanguageClient: Started initializing a new COBOL language client with these initialization options: {"enableUnreachableCodeWarnings":true,"publishDiagnostics":true,"includeFileExtensions":[".cpy",".copybook",".copy",".cobcopy"],"compilerOptions":""}.
2024-02-04T15:49:41.714+01:00 DEBUG: AbstractZLanguageClient.startLanguageClient: Language client for COBOL started.
```
![image](https://github.com/IBM/zopeneditor-about/assets/63861965/feaaceb7-6d8c-442f-962b-4c38e10b9ba1)

The expression `${execPath}\..\data\JAVA\bin\java.exe` should be resolved in `X:\installation_folder\Code.exe\..\data\JAVA\bin\java.exe`, what doesn't work...

However, in a "Windows Terminal" the syntax is accepted:
```
X:\>X:\installation_folder\Code.exe\..\data\JAVA\bin\java.exe --version
openjdk 11.0.16.1 2022-08-12
IBM Semeru Runtime Open Edition 11.0.16.1 (build 11.0.16.1+1)
Eclipse OpenJ9 VM 11.0.16.1 (build openj9-0.33.1, JRE 11 Windows 11 amd64-64-Bit Compressed References 20220818_429 (JIT enabled, AOT enabled)
OpenJ9   - 1d9d16830
OMR      - b58aa2708
JCL      - 84b84a0409 based on jdk-11.0.16.1+1)
```

In conclusion, my analysis is that JavaFinder manages to execute the JDK indicated in the `zopeneditor.JAVA_HOME` parameter of User `settings.json`, but ZOE cannot execute it...

If the JavaFinder module used is the one available here, https://github.com/HanSolo/javafinder, we see that it resolves the access path:
https://github.com/HanSolo/javafinder/blob/709f5825863518b0ffbe24c70de205edd0900d40/src/main/java/eu/hansolo/javafinder/Finder.java#L114 
See: https://docs.oracle.com/javase/8/docs/api/java/nio/file/Paths.html#get-java.lang.String-java.lang.String...-

Would it be possible for ZOE to do the same Java executable path resolution as JavaFinder?

Failing that, do you know of a method to declare an access path relative to the `zopeneditor.JAVA_HOME` parameter?

Note: same problem with [IBM Db2 for z/OS Developer extension](https://github.com/IBM/db2forzosdeveloperextension-about), probably same reason.

Thanks.

**Comment by FALLAI-Denis** [comment]
PS: 

I found a workaround by creating a launch file (.bat or .cmd) that declares the `JAVA_HOME` environment variable with the path to the folder containing the portable installation's Java runtime and resetting to default the `zopeneditor.JAVA_HOME` parameter:

```
@echo off
set PORTABLE_INST=%~dp0installation_folder
set JAVA_HOME=%PORTABLE_INST%\data\JAVA
start "Portable Installation" /D "%PORTABLE_INST%" "%PORTABLE_INST%\Code.exe" 
```

This work also for other extensions needing a Java runtime.

It is then imperative to start the potable installation of VS Code with the created command file.

**Comment by phaumer** [comment]
@FALLAI-Denis thanks for the detailed (and as always) correct analysis. We currently find Java based on the user setting, the JAVA_HOME variable, if Java is in the PATH, plus we even look in the most common installation locations for Windows and Mac. The fact that the command starting the language server does not resolve environment variables is something we can look into improving.
